### PR TITLE
User Operator: Authorization CRD changes

### PR DESCRIPTION
**Pull request by scholzj**
### Type of change

- ~~Bugfix~~
- Enhancement / new feature
- ~~Refactoring~~
- ~~Documentation~~

### Description

This PR delivers the changes to the `KafkaUser` CRD which are related to the authorisation. I will see whether the implementation causes some changes. But basically I tried to break it into the simplest possible way, so it basically consists of 4 pieces:
- type
- resource
- host
- operation

Type could be either allow or deny. Resource specifies either topic or group or cluster. Topic and group have a name and pattern type (prefix wildcards).

Compared to the interface from Kafka there are several notable changes:
- Only one resource can be specified per rule
- Only one host can be specified per rule
- Only one operation can be specified per rule
- No shortcuts for producer or creator

I thought this could simplify the internal design and maybe make it easier to read. But the price is much more lines than you would need commands with `kafka_acls.sh`. Do you think my thinking is valid? Or should we for example allow list of resources, operations or hosts?

Here is an example of how the actual YAML might look like. There are defaults for hosts and type, so they do not have to be always specified.

```yaml
apiVersion: kafka.strimzi.io/v1alpha1
kind: KafkaUser
metadata:
  name: my-user
spec:
  authentication:
    type: tls
  authorization:
    type: simple
    acls:
      - type: allow
        resource:
          type: topic
          name: my-topic
          patternType: prefix
        operation: Read
      - resource:
          type: topic
          name: my
          patternType: prefix
        operation: Read
        host: 127.0.0.1
      - resource:
          type: group
          name: my-group
        operation: Read
        host: 127.0.0.1
      - resource:
          type: cluster
        operation: Read
```

### Checklist

- [x] Make sure all tests pass

### CODE REVIEW

**Comment by ppatierno** [comment]
Personally, I don't like so much the possibility in the `kafka_acls.sh` script to allow mixing operations on multiple resources for multiple principals in one command, so your proposal makes all more clear.

I think that having only one operation per resource is reasonable.
In data-streaming it's quite difficult that an application reads/writes from/to the same topic. In general, it reads from a topic and writes to a different one.
Of course, if the same user is used for ACLs in two different applications (one reading and the other one writing on the same topic), my consideration isn't valid anymore right?

Regarding the patternType field. Just an idea but maybe it makes more difficult the internal design; could we have "name" and "prefix" as two different fields which cannot be used at the same time?
It means that if you want to address a specific topic name my-topic you have to use "name: my-topic"; if you want to address all topics starting with "temp" you have to use "prefix: temp".
Instead of having one more field as "patternType".

Maybe just having hosts list should not making too much complex the design?

**Comment by scholzj** [comment]
> Regarding the patternType field. Just an idea but maybe it makes more difficult the internal design; could we have "name" and "prefix" as two different fields which cannot be used at the same time?
It means that if you want to address a specific topic name my-topic you have to use "name: my-topic"; if you want to address all topics starting with "temp" you have to use "prefix: temp".
Instead of having one more field as "patternType".

I was thinking about that and had it in my original design. But I think it is a bad pattern. You cannot enforce it at the CRD level. So you end up allowing the user to specify both or none of them. And then you have to deal with it in the application by printing warnings. The design in this PR is IMHO much more clear and leaves no space for mistakes.

**Comment by ppatierno** [comment]
So I would not say that it's a bad pattern it's just that the CRD doesn't support having two fields which are exclusive each other :)
Agree that providing warnings is not the best way.

**Comment by scholzj** [comment]
@ppatierno Yeah ... if we could set some fields as mutually exclusive I would be all for it :-/

File: api/src/main/java/io/strimzi/api/kafka/model/AclOperation.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+            case "AlterConfigs":
+                return ALTERCONFIGS;
+            case "DescribeConfigs":
+                return DESCRIBECONFIGS;
+            case "IdempotentWrite":
+                return IDEMPOTENTWRITE;
+            case "All":
+                return ALL;
+
+
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/AclOperation.java]: too many blank lines :-)

File: api/src/main/java/io/strimzi/api/kafka/model/AclRuleGroupResource.java [hunk 2]
```diff
@@ -0,0 +1,60 @@
...
+    @Description("Must be `" + TYPE_GROUP + "`")
+    @Override
+    public String getType() {
+        return TYPE_GROUP;
+    }
+
+    @Description("Describes the pattern used in the resource field. " +
+            "The supported types are `literal` and `prefix`. " +
+            "With `literal` pattern type, the resource field will be used as a definition of a full topic name. " +
+            "With `prefix` pattern type, the resoruce name will be used only as a prefix. " +
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/AclRuleGroupResource.java]: typo "resource"

File: api/src/main/java/io/strimzi/api/kafka/model/AclRuleTopicResource.java [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    @Description("Must be `" + TYPE_TOPIC + "`")
+    @Override
+    public String getType() {
+        return TYPE_TOPIC;
+    }
+
+    @Description("Describes the pattern used in the resource field. " +
+            "The supported types are `literal` and `prefix`. " +
+            "With `literal` pattern type, the resource field will be used as a definition of a full topic name. " +
+            "With `prefix` pattern type, the resoruce name will be used only as a prefix. " +
```
Comment (ppatierno) [inline api/src/main/java/io/strimzi/api/kafka/model/AclRuleTopicResource.java]: typo "resource" (I know ... copy/paste ;))

**Comment by scholzj** [comment]
@ppatierno What additional tests would you like to see? I think it is tested as all other CRDs. At the end its just JSON encoding and decoding without a real schema, there is no real functionality.

**Comment by ppatierno** [comment]
Not so much, at least one test with missing required fields? i.e. "operation" and/or "resource" not specified under an ACL rule? In any case, I'm going to approve this PR ;)

**Comment by scholzj** [comment]
@ppatierno This is the same as with all other CRDs. We also don't test there every single required field. This should be the point of the generator that you can trust that it generates the the CRDs as they should be..

**Pull request merged**